### PR TITLE
Add a Scoreboard which stores data in localStorage

### PR DESCRIPTION
**Pull request by sohammondal**
- Render scoreboard as a table only if at least one move is saved
- Take user name as input after game finishes
- Show Thank you after a user saves his / her moves
- The scoreboard is updated after user refreshes the page

### CODE REVIEW

**Comment by itaditya** [comment]
@sohammondal the JS part is fine but design part is not. In my local I have improved the design of your code. Can you grant me edit permissions in your PR, so that I can push my commit to your PR also. Refer this to understand how this is done. https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

File: src/Game.js [hunk 1]
```diff
@@ -40,6 +42,50 @@ class Game extends Component {
     return spritesPos;
   }
 
+  /*
```
Comment (itaditya) [inline src/Game.js]: Can you remove these comments. There are various ways on GitHub to know who contributed what code. There's Contributions graphs, `git blame`.

Plus I'll be adding every contributors name in the README.

Comments like these obfuscate the actual logic.

File: src/Game.js [hunk 2]
```diff
@@ -40,6 +42,50 @@ class Game extends Component {
     return spritesPos;
   }
 
+  /* 
+    Takes username input to save moves
+     ** Contributed by Soham Mondal <contact@sohammondal.com> **
+  */
+  userNameInputForm() {
```
Comment (itaditya) [inline src/Game.js]: Rename this function to `renderUsernameForm`

File: src/Game.js [hunk 3]
```diff
@@ -40,6 +42,50 @@ class Game extends Component {
...
+  saveMoves(name, moves) {
+    // console.log(moves);
+    const user = {
+      name,
+      moves
+    };
+    let savedUserMoves = localStorage.getItem("userMoves");
+    if (savedUserMoves) {
+      savedUserMoves = JSON.parse(savedUserMoves);
+      // console.log(savedUserMoves);
```
Comment (itaditya) [inline src/Game.js]: remove this console log

File: src/Game.js [hunk 4]
```diff
@@ -40,6 +42,50 @@ class Game extends Component {
...
+    Save moves to localStorage 
+     ** Contributed by Soham Mondal <contact@sohammondal.com> **
+  */
+  saveMoves(name, moves) {
+    // console.log(moves);
+    const user = {
+      name,
+      moves
+    };
+    let savedUserMoves = localStorage.getItem("userMoves");
```
Comment (itaditya) [inline src/Game.js]: use single quote

File: src/Game.js [hunk 5]
```diff
@@ -40,6 +42,50 @@ class Game extends Component {
...
+    if (savedUserMoves) {
+      savedUserMoves = JSON.parse(savedUserMoves);
+      // console.log(savedUserMoves);
+      savedUserMoves.push(user);
+      localStorage.setItem("userMoves", JSON.stringify(savedUserMoves));
```
Comment (itaditya) [inline src/Game.js]: use single quote

File: src/Game.js [hunk 6]
```diff
@@ -49,6 +95,7 @@ class Game extends Component {
     const hasFinished = this.remainingSprites === 0;
 
     if (!this.state.hasFinished && hasFinished) {
+      // this.saveMoves(this.moves);
```
Comment (itaditya) [inline src/Game.js]: remove comment

File: src/ScoreBoard.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  getUserMovesFromLocalStorage() {
+    let savedUserMoves = localStorage.getItem("userMoves");
+    if (savedUserMoves) {
+      savedUserMoves = JSON.parse(savedUserMoves);
+      savedUserMoves.sort((user1, user2) => user1.moves - user2.moves);
+      this.setState({ userMoves: savedUserMoves });
+    }
+  }
+
+  renderScoreboard(userMovesArray) {
```
Comment (itaditya) [inline src/ScoreBoard.js]: instead of passing state data to render function via params you can just use the state in the function itself.

Something like

```js
renderScoreboard() {
  const { userMoves } = this.state;
  // now user userMoves instead of userMovesArray.
}
```

File: src/scoreboard.css [hunk 8]
```diff
@@ -0,0 +1,11 @@
+.scoreboard {
+  position: absolute;
+  color: white;
+  /* border: 1px solid white; */
```
Comment (itaditya) [inline src/scoreboard.css]: remove comment

File: src/styles.css [hunk 9]
```diff
@@ -23,7 +23,7 @@ body {
   outline: 0;
   border-collapse: collapse;
   background-color: rgba(0, 0, 0, 0.6);
-  background-image: url('/images/board-background.png');
+  background-image: url("/images/board-background.png");
```
Comment (itaditya) [inline src/styles.css]: revert this to single quote.

File: src/styles.css [hunk 10]
```diff
@@ -38,20 +38,20 @@ body {
 }
 
 .board-cell.sprite-1 {
-  background-image: url('/images/sweet-1.png');
+  background-image: url("/images/sweet-1.png");
```
Comment (itaditya) [inline src/styles.css]: revert all to single quotes

File: src/Game.js [hunk 11]
```diff
@@ -40,6 +42,50 @@ class Game extends Component {
...
+      </span>
+    );
+  }
+
+  /* 
+    Save moves to localStorage 
+     ** Contributed by Soham Mondal <contact@sohammondal.com> **
+  */
+  saveMoves(name, moves) {
+    // console.log(moves);
```
Comment (itaditya) [inline src/Game.js]: remove console log

**Comment by itaditya** [comment]
@sohammondal left some Change Requests as well. Your work is superb. Keep going!!

File: src/Game.js (hunk 1 above)
Comment (sohammondal) [reply src/Game.js]: Yes, you're absolutely right. I will remove all the unnecessary comments.

File: src/Game.js (hunk 2 above)
Comment (sohammondal) [reply src/Game.js]: Done

File: src/Game.js (hunk 4 above)
Comment (sohammondal) [reply src/Game.js]: I actually have Prettier setup with VS Code which changes all single quotes to double-quotes.

File: src/Game.js (hunk 5 above)
Comment (sohammondal) [reply src/Game.js]: Done

**Comment by sohammondal** [comment]
> 
> 
> @sohammondal left some Change Requests as well. Your work is superb. Keep going!!

Thank you for the encouragement. 😄  I have updated the code as per the change requests. Please review.

**Comment by itaditya** [comment]
can you give me access to add commits in your PR https://github.com/itaditya/trick-or-treat-game/pull/24#issuecomment-537149823

**Comment by itaditya** [comment]
There is one bug in your code. You are sending `name` of team. Actually name can be 'My Team' but slug would be `my-team` and every further GitHub API uses the team slug to get that team's data.

The `name` field can be used in the dropdown as the label. That means in dropdown user will see the list of names of teams. When they click on one item, we will find the slug corresponding to that item and then go ahead.

So basically we need both `name` and `slug` from API, so you should API res->

```json
{
  "teamsList": [
    {
      "name": "Some Team",
      "slug": "someteam"
    },
    {
      "name": "My Team",
      "slug": "my-team"
    }
  ]
}

```

**Comment by sohammondal** [comment]
> 
> 
> can you give me access to add commits in your PR [#24 (comment)](https://github.com/itaditya/trick-or-treat-game/pull/24#issuecomment-537149823)

Hey, I have already [✔️] the checkbox.  
image

You should be good to go. Can you please check?

File: src/Game.js [hunk 12]
```diff
@@ -194,6 +231,13 @@ class Game extends Component {
             &nbsp; moves
             <br />
             <br />
+            {this.state.isMovesSaved ? (
+              <span>Thanks, {this.state.username}! Your moves are saved.</span>
+            ) : (
+                this.userNameInputForm()
```
Comment (itaditya) [inline src/Game.js]: rename to `renderUsernameForm`

**Comment by itaditya** [comment]
@sohammondal congratulations on getting your PR merged. This was a significant feature and I'm so happy it got landed.

Congrats

**Comment by itaditya** [comment]
@sohammondal congratulations on getting your PR merged. This was a significant feature and I'm so happy it got landed.

Congrats

**Comment by sohammondal** [comment]
@itaditya Thanks for merging the PR. It was a very good learning experience for me.
